src/cider/nrepl/middleware/track_state.clj

```diff
@@ -0,0 +1,55 @@
+(ns cider.nrepl.middleware.track-state
+  "State tracker for client sessions."
+  {:author "Artur Malabarba"}
+  (:require [cider.nrepl.middleware.util.cljs :as cljs]
+            [clojure.tools.nrepl.middleware :refer [set-descriptor!]]
+            [clojure.tools.nrepl.middleware.interruptible-eval :refer [*msg*]]
+            [clojure.tools.nrepl.misc :refer [response-for]]
+            [clojure.tools.nrepl.transport :as transport])
+  (:import clojure.tools.nrepl.transport.Transport))
+
+
+(def messages (atom {}))
+
+(defn notify-state
+  "Reply to the message associated with `session`."
+  [session]
+  (when-let [msg (@messages session)]
+    (->> (if (cljs/grab-cljs-env *msg*) :cljs :clj)
+         (response-for msg :repl-type)
+         (transport/send (:transport msg)))))
+
+(defn make-transport
+  [{:keys [^Transport transport session]}]
+  (reify Transport
+    (recv [this] (.recv transport))
+    (recv [this timeout] (.recv transport timeout))
+    (send [this response]
+      (.send transport response)
+      (when (:done (:status response))
```

Comment:
Minor style nit - this could be `(get-in response [:status :done])`, but I would go so far as to do `(contains? (:status response) :done)` to make it clear that we're checking for the presence of `:done` in a set.